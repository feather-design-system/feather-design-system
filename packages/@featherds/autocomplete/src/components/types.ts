import { Ref, PropType } from "vue";
import HighlightProps from "./Highlight/HighlightProps";
import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
interface IAutocompleteItem extends IAutocompleteItemType {
  _new?: string;
  _text: string;
  _pre?: IAutocompleteChipIcon;
}
interface IAutocompleteItemType {
  [k: string]: unknown;
}

interface IAutocompleteGridColumn {
  prop: keyof IAutocompleteItemType;
  title: string;
  align?: "right" | "left" | "center";
}

interface IAutocompleteResultRender {
  reset: () => void;
  handleKeyPress: (
    e: KeyboardEvent,
    internalResults: IAutocompleteItemType[]
  ) => boolean;
  active: {
    row: number;
    col?: number;
  };
  first: () => void;
}

export enum AutocompleteTypes {
  multi = "multi",
  single = "single",
}
interface IStrategyOptions {
  selectionLimit: Ref<number>;
  selectionLimitReached: Ref<boolean>;
  modelValue: Ref<IAutocompleteItemType | IAutocompleteItemType[]>;
  textProp: Ref<string>;
  input: Ref<HTMLInputElement>;
  query: Ref<string>;
  emitSearch: () => void;
  forceCloseResults: Ref<boolean>;
  allowNew: Ref<boolean>;
  internalResults: Ref<IAutocompleteItemType[]>;
}

interface IAutocompleteChipIcon {
  icon: unknown;
  title: string;
}

export type {
  IStrategyOptions,
  IAutocompleteGridColumn,
  IAutocompleteItemType,
  IAutocompleteItem,
  IAutocompleteChipIcon,
  IAutocompleteResultRender,
};

export const LABELS = {
  noResults: "No results found",
  minChar: "Enter ${min} characters to search",
  clear: "Clear value",
  selectionLimit: "Selection limit reached",
  new: "New",
  remove: "Remove",
};
export const props = {
  id: {
    type: String,
  },
  modelValue: {
    type: [Array, Object] as PropType<
      IAutocompleteItemType[] | IAutocompleteItemType
    >,
  },
  results: {
    type: Array as PropType<IAutocompleteItemType[]>,
    default: () => [],
  },
  textProp: {
    type: String as unknown as PropType<keyof IAutocompleteItemType>,
    default: "_text",
  },
  loading: {
    type: Boolean,
    default: false,
  },
  minChar: {
    type: Number,
    default: 0,
  },
  allowNew: {
    type: Boolean,
    default: false,
  },
  selectionLimit: {
    type: Number,
  },
  newMatcher: {
    type: Function,
    default: (
      item: IAutocompleteItemType,
      query: string,
      comp: { textProp: string }
    ) => {
      return (
        (item[comp.textProp] as string).toString().toLowerCase() ===
        query.toLowerCase()
      );
    },
  },
  type: {
    type: String as PropType<keyof typeof AutocompleteTypes>,
    required: true,
    validator: (v: keyof typeof AutocompleteTypes) => {
      // The value must match either
      return !!AutocompleteTypes[v];
    },
  },
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default() {
      return LABELS;
    },
  },
  gridConfig: {
    type: Array as PropType<IAutocompleteGridColumn[]>,
  },
  schema: {
    type: Object,
    required: false,
  },
  ...HighlightProps,
  ...InputSubTextProps,
  ...InputWrapperProps,
} as const;

export const emits = {
  "update:modelValue": (
    value: IAutocompleteItemType | IAutocompleteItemType[] | undefined
  ) => true,
  new: (value: string) => true,
  search: (value: string) => true,
};
