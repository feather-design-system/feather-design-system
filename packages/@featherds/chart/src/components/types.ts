export type ShirtSize = "xs" | "sm" | "md" | "lg" | "xl";

export type FeatherChartType = "bar";

export type FeatherBarChartData = {
  labels: Array<string>;
  datasets: Array<{
    label: string;
    data: Array<string | number>;
    backgroundColor?: string;
  }>;
};

export type FeatherBarChartOptions = {
  responsive?: boolean;
  maintainAspectRation?: boolean;
};

export type FeatherD3Dimensions = {
  height: number;
  width: number;
  margin: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  }
};

export type FeatherD3Options = {
  units: string;
  class?: "color1" | "color2";
};


// Data can be any array of objects returned from an API or local JSON
export type FeatherD3Data = Array<unknown>;

export interface ICoordinateAxes {
  x: string,
  y: string
}
