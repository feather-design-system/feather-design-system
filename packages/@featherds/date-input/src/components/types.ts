export interface IDateDisabledConfig extends IDateRange {
  dates?: (Date | number)[];
  ranges?: IDateRange[];
  days?: number[];
  daysOfMonth?: number[];
  customPredictor?: (date: Date | number) => boolean;
}
interface IDateRange {
  to?: Date | number;
  from?: Date | number;
}
export const LABELS = {
  day: "Day",
  month: "Month",
  year: "Year",
  prevMonth: "Previous month",
  nextMonth: "Next month",
  selectMonth: "Select month",
  selectYear: "Select year",
  inCalendar: "In calendar",
  calendar: "Click for calendar",
  clear: "Clear date",
  menu: "Use arrow keys to navigate dates. Page down and page up will navigate by month. Shift page down and shift page up will navigate by year. Press escape to exit the calendar.",
};
