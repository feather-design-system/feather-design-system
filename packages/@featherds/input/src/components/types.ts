import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
import { PropType } from "vue";
export const props = {
  type: {
    type: String,
    default: "text",
  },
  modelValue: {
    type: [String, Number] as PropType<string | number>,
  },
  maxlength: {
    type: Number,
    required: false,
    default: 0,
  },
  schema: {
    type: Object as PropType<Record<string, any>>,
    required: false,
  },
  id: {
    type: String,
    required: false,
  },
  ...InputWrapperProps,
  ...InputSubTextProps,
} as const;

export const emits = {
  "update:modelValue": (value: string | undefined | number) => true,
};
