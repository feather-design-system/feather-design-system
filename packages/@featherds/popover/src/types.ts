// Note: introduced new types
export type Tooltip = {
  height: number;
  width: number;
};

export type Trigger = {
  height: number;
  width: number;
  left: number;
  top: number;
};

export enum PopoverPlacement {
  top = "top",
  bottom = "bottom",
  left = "left",
  right = "right",
}

export enum PointerAlignment {
  center = "center",
  left = "left",
  right = "right",
}
