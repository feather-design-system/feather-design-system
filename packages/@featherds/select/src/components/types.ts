export interface ISelectItem extends ISelectItemType {
  _text: string;
}

export interface ISelectItemType {
  [k: string]: unknown;
}
