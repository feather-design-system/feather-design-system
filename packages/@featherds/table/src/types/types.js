export const ALIGNMENT = {
  LEFT: "left",
  RIGHT: "right",
  CENTER: "center",
};

export const SORT = {
  ASCENDING: "asc",
  DESCENDING: "desc",
  NONE: "none",
};
