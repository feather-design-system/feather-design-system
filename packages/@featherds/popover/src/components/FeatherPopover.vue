<template>
  <slot name="trigger" :attrs="attrs" :on="listeners"> </slot>
  <Transition :css="animate">
    <div
      class="feather-popover-container"
      v-if="show"
      ref="popover"
      :style="{ left: left, top: top }"
      :class="[alignmentClass, placementClass]"
    >
      <div class="popover" :id="popoverID"><slot /></div>
      <Pointer class="popover-pointer" />
    </div>
  </Transition>
</template>
<script lang="ts">
import { useScroll } from "@featherds/composables/events/Scroll";
import { getSafeId } from "@featherds/utils/id";
import { Code } from "@featherds/utils/keys";
import Pointer from "./PopoverPointer.vue";
import {
  defineComponent,
  toRef,
  computed,
  ref,
  Ref,
  PropType,
  nextTick,
  provide,
  onUnmounted,
  watch,
  onMounted,
} from "vue";
import { PointerAlignment, PopoverPlacement } from "../types";
import { calculatePlacement } from "./Placement";
import { calculateAlignment } from "./Alignment";
import { useResize } from "@featherds/composables/events/Resize";
import { useOutsideClick } from "@featherds/composables/events/OutsideClick";
export const props = {
  placement: {
    type: String as PropType<PopoverPlacement>,
    default: () => PopoverPlacement.top,
  },
  pointerAlignment: {
    type: String as PropType<PointerAlignment>,
    default: () => PointerAlignment.center,
  },
} as const;
export default defineComponent({
  props,
  setup(props) {
    const show = ref(false); //show will trigger animate when positioning is ready
    const animate = ref(false); //animate triggers the visual
    const popover = ref() as Ref<HTMLElement>;
    const triggerID = getSafeId("feather-popover-trigger");
    const popoverID = getSafeId("feather-popover");
    const idAttr = "data-feather-popover";
    provide("feather-has-popover", true);

    const handleKeypress = (e: KeyboardEvent) => {
      if (e.code === Code.ESCAPE) {
        e.preventDefault();
        hidePopover(true);
      }
      if (e.code === Code.ENTER || e.code === Code.SPACE) {
        e.preventDefault();
        togglePopover();
      }
    };
    const togglePopover = () => {
      if (show.value) {
        hidePopover();
      } else {
        showPopover();
      }
    };
    const attrs = computed(() => {
      return {
        [idAttr]: triggerID,
        "aria-controls": popoverID,
        "aria-expanded": show.value,
      };
    });
    const listeners = {
      click: togglePopover,
      keydown: handleKeypress,
    };

    onMounted(() => (windowRef.value = window));
    onUnmounted(() => hidePopover(true));
    const getPopover = () => document.getElementById(popoverID) as HTMLElement;
    const showPopover = () => {
      animate.value = false;
      show.value = true; //show it but dont animate
      nextTick(() => {
        const popover = getPopover();
        positionPopover(popover);
        show.value = false;
        nextTick(() => {
          animate.value = true;
          show.value = true; //show and animate now :D
        });
      });
    };
    const hidePopover = (force = false) => {
      //reset transform and alignment class
      left.value = "";
      top.value = "";
      alignmentClass.value = "";
      placementSelected.value = "";
      show.value = false;
      if (force) {
        animate.value = false;
      }
    };
    const windowRef = ref() as Ref<Window>;
    const scrollActivate = useScroll(windowRef, () => {
      positionPopover(getPopover());
    });
    const resizeActivate = useResize(() => {
      hidePopover(true);
    });
    const outsideClickActivate = useOutsideClick(
      computed(() => {
        return [
          popover.value,
          document.querySelector(`[${idAttr}=${triggerID}]`),
        ];
      }) as Ref<HTMLElement[]>,
      () => {
        hidePopover(true);
      },
      { click: true }
    );

    watch(show, (v) => {
      scrollActivate.value = v;
      resizeActivate.value = v;
      outsideClickActivate.value = v;
    });

    const placementProp = toRef(props, "placement");
    const alignmentProp = toRef(props, "pointerAlignment");
    const arrowHeight = 18;
    const arrowOffset = 31;
    const top = ref("");
    const left = ref("");
    const alignmentClass = ref("");
    const placementSelected = ref("");
    const placementClass = computed(() => {
      return placementSelected.value ? "p-" + placementSelected.value : false;
    });

    const positionPopover = (popoverElement: HTMLElement) => {
      const triggerElement = document.querySelector(`[${idAttr}=${triggerID}]`);

      if (!triggerElement) {
        console.log("trigger not found");
        return;
      }
      nextTick(() => {
        const triggerBox = triggerElement.getBoundingClientRect();
        const popoverBox = popoverElement.getBoundingClientRect();
        const placement = calculatePlacement(
          triggerBox,
          popoverBox,
          placementProp.value,
          arrowHeight
        ) as PopoverPlacement; // ts(18048)
        const alignment = calculateAlignment(
          placement,
          triggerBox,
          popoverBox,
          alignmentProp.value,
          arrowOffset
        ) as PointerAlignment; // ts(18048)
        //set the classes so arrow can be positioned
        alignmentClass.value = alignment.toString();
        placementSelected.value = placement.toString();

        //calculate the transform based off alignment and placement
        let topCal = 0;
        let leftCal = 0;
        if (
          placement === PopoverPlacement.left ||
          placement === PopoverPlacement.right
        ) {
          //align center points of trigger and popover
          //so they can be side by side
          const triggerBoxCenter = triggerBox.top + triggerBox.height / 2;
          //find the left spacing
          if (placement === PopoverPlacement.left) {
            leftCal = triggerBox.left - popoverBox.width - arrowHeight;
          }
          if (placement === PopoverPlacement.right) {
            leftCal = triggerBox.right;
          }

          switch (alignment) {
            case PointerAlignment.center:
              //align centers
              topCal = triggerBoxCenter - popoverBox.height / 2;
              break;
            case PointerAlignment.left:
              //align left but take into consideration arrow offset
              topCal = triggerBoxCenter - arrowOffset;
              break;
            case PointerAlignment.right:
              //align to the right and allow space for arrow offset
              topCal = triggerBoxCenter - popoverBox.height + arrowOffset;
              break;
          }
        }

        if (
          placement === PopoverPlacement.top ||
          placement === PopoverPlacement.bottom
        ) {
          //get vertical position
          //top must take into cosideration arrow
          topCal = triggerBox.top - popoverBox.height - arrowHeight;
          if (placement === PopoverPlacement.bottom) {
            topCal = triggerBox.bottom;
          }
          const triggerBoxCenter = triggerBox.left + triggerBox.width / 2;
          switch (alignment) {
            case PointerAlignment.center:
              //align centers
              leftCal = triggerBoxCenter - popoverBox.width / 2;
              break;
            case PointerAlignment.left:
              //align left but take into consideration arrow offset
              leftCal = triggerBoxCenter - arrowOffset;
              break;
            case PointerAlignment.right:
              //align to the right and allow space for arrow offset
              leftCal = triggerBoxCenter - popoverBox.width + arrowOffset;
              break;
          }
        }

        top.value = topCal.toString() + "px";
        left.value = leftCal.toString() + "px";
        return;
      });
    };

    return {
      attrs,
      listeners,
      show,
      animate,
      alignmentClass,
      placementClass,
      top,
      left,
      popoverID,
      popover,
    };
  },
  components: {
    Pointer,
  },
});
</script>

<style lang="scss" scoped>
@use "sass:math";
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/mixins/typography";
@import "@featherds/styles/mixins/elevation";
$pWidth: 33px;
$pHeight: 18px;
$expectedWidth: 18px;
$offset: 15px;
$verticalOffset: $offset + (math.div($pWidth, 2) - math.div($pHeight, 2));

.feather-popover-container {
  position: fixed;
  top: 0;
  left: 0;
  z-index: var($zindex-popover);
  .popover {
    @include body-small;
    color: var($secondary-text-on-surface);
    background-color: var($elevation-background-2);
    border: 1px solid var($border-on-surface);
    padding: 1.25rem;
    border-radius: 4px;
    max-width: 16rem;
  }
  .popover-pointer {
    position: absolute;
    transform-origin: center center;
  }

  &.p-bottom {
    .popover {
      margin-top: $expectedWidth;
    }
    .popover-pointer {
      top: 4px;
      transform: rotate(180deg);
    }
    &.center .popover-pointer {
      left: 50%;
      transform: translateX(-50%) rotate(180deg);
    }
    &.right .popover-pointer {
      right: $offset;
    }
    &.left .popover-pointer {
      left: $offset;
    }
  }
  &.p-top {
    .popover-pointer {
      top: calc(100% - 3px);
    }
    &.center .popover-pointer {
      left: 50%;
      transform: translateX(-50%);
    }
    &.right .popover-pointer {
      right: $offset;
    }
    &.left .popover-pointer {
      left: $offset;
    }
  }

  &.p-left {
    .popover {
      margin-right: $expectedWidth;
    }
    .popover-pointer {
      right: -4px;

      transform: rotate(-90deg);
    }

    &.center .popover-pointer {
      top: calc(50% - ($pHeight/2));
    }
    &.right .popover-pointer {
      bottom: $verticalOffset;
    }
    &.left .popover-pointer {
      top: $verticalOffset;
    }
  }
  &.p-right {
    .popover {
      margin-left: $expectedWidth;
    }
    .popover-pointer {
      left: -4px;
      transform: rotate(90deg);
    }
    &.center .popover-pointer {
      top: calc(50% - ($pHeight/2));
    }
    &.right .popover-pointer {
      bottom: $verticalOffset;
    }
    &.left .popover-pointer {
      top: $verticalOffset;
    }
  }
}

.v-enter-active,
.v-leave-active {
  transition-property: opacity, transform;
  transition-duration: 280ms;
  transition-timing-function: ease-in-out;
}

// .v-enter-from,
// .v-leave-to {
//   opacity: 0;
//   transform: translateY(4px);
// }

.p-top {
  &.v-enter-from,
  &.v-leave-to {
    opacity: 0;
    transform: translateY(4px);
  }
}

.p-bottom {
  &.v-enter-from,
  &.v-leave-to {
    opacity: 0;
    transform: translateY(-4px);
  }
}
.p-left {
  &.v-enter-from,
  &.v-leave-to {
    opacity: 0;
    transform: translateX(4px);
  }
}

.p-right {
  &.v-enter-from,
  &.v-leave-to {
    opacity: 0;
    transform: translateX(-4px);
  }
}

.v-enter-to,
.v-leave-from {
  transform: translateY(0) translateX(0);
}
</style>
