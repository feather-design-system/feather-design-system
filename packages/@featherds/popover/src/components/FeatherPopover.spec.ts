import { mount } from "@vue/test-utils";
import { nextTick } from "vue";
import FeatherPopover from "./FeatherPopover.vue";
import * as id from "@featherds/utils/id";
import { PopoverPlacement } from "../types";
import { Code } from "@featherds/utils/keys";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);
const getTrigger = () =>
  `<template #trigger="params">
  <button v-bind="params.attrs" v-on="params.on">popover BUTTON</button>
  </template>
`;

const getDefault = () =>
  `<p>Hello from a popover</p>
`;
const getWrapper = (placement = PopoverPlacement.top) =>
  mount(FeatherPopover, {
    props: {
      title: "hello world",
      placement,
    },
    slots: {
      default: getDefault(),
      trigger: getTrigger(),
    },
    attachTo: "body",
  });
const click = async (wrapper: ReturnType<typeof getWrapper>) => {
  await wrapper.find("button").trigger("click");
  await nextTick();
};

describe("FeatherPopover", () => {
  it("should not show a popover until click", () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should show a popover on click", async () => {
    jest.useFakeTimers();
    const wrapper = getWrapper();
    await click(wrapper);
    jest.runAllTimers();
    await nextTick();
    await nextTick();
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should hide popover on escape", async () => {
    jest.useFakeTimers();
    const wrapper = getWrapper();
    await click(wrapper);
    jest.runAllTimers();
    await nextTick();
    await nextTick();
    await nextTick();

    const button = wrapper.find("button");
    await button.trigger("keydown", { code: Code.ESCAPE });

    expect(wrapper.element).toMatchSnapshot();
  });
});
