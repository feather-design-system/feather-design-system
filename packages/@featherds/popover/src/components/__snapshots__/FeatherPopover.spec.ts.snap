// Vitest Snapshot v1

exports[`FeatherPopover > should hide popover on escape 1`] = `
<div>
  
  
  <button
    aria-controls="feather-popover"
    aria-expanded="false"
    data-feather-popover="feather-popover-trigger"
  >
    popover BUTTON
  </button>
  
  <transition-stub
    appear="false"
    css="false"
    persisted="false"
  >
    <!--v-if-->
  </transition-stub>
  
</div>
`;

exports[`FeatherPopover > should not show a popover until click 1`] = `
<div>
  
  
  <button
    aria-controls="feather-popover"
    aria-expanded="false"
    data-feather-popover="feather-popover-trigger"
  >
    popover BUTTON
  </button>
  
  <transition-stub
    appear="false"
    css="false"
    persisted="false"
  >
    <!--v-if-->
  </transition-stub>
  
</div>
`;

exports[`FeatherPopover > should show a popover on click 1`] = `
<div>
  
  
  <button
    aria-controls="feather-popover"
    aria-expanded="false"
    data-feather-popover="feather-popover-trigger"
  >
    popover BUTTON
  </button>
  
  <transition-stub
    appear="false"
    css="false"
    persisted="false"
  >
    <!--v-if-->
  </transition-stub>
  
</div>
`;
